### DISCUSSION TITLE
SF & SAHI

### DISCUSSION THREAD

**Discussion by cosminlovin7** [ultralytics/ultralytics#23515] [Q&A] [answered]
Hello, i am considering improving the small object detection using Slicing Aided Fine-tuning (SF) and Slicing Aided Hyper Inference (SAHI).

I have a few questions:
1. In order to apply Slicing Aided Fine-tuning (SF), it is pretty clear that the training set has to be augmented by slicing each image in tiles with some overlap between tiles. However, when validating, should the validation set be also augmented (in the same way we augmented the training set), or should we directly use SAHI approach for validation.
2. Considering that our images are 4K resolution, with very small objects in the pictures, when tiling at 640x640 px, we may end up with only one or two tiles containing the objects, and the rest of ~30 tiles are just background. I assume it may not be ok to include all negative tiles in the training set, especially in this ratio 1:30. I see in the SAHI implementation that there is a flag "ignore_negative_samples" which is False by default. I guess it may not include the negative tiles when set to True. For Slicing Aided Fine-tuning (SF), should we set this flag to True, should we include only a few negative samples, in a smaller ratio, maybe like 1:3 or 1:5, or should we just include all negative samples?

Thank you very much!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cosminlovin7, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started quickly, and an Ultralytics engineer will also assist soon 🙂  
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report (e.g., unexpected metrics, crashes, or incorrect slice/merge behavior), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (which it sounds like, regarding SF/SAHI workflows for small objects), please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).  
In particular, sharing a snippet of your tiling setup (tile size/overlap), your validation approach, and how many “negative/background” tiles you’re generating would help the team give targeted guidance 🧩

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
For SF, I’d keep your validation split in its original 4K form and evaluate using the same end-to-end SAHI sliced-inference + merge pipeline you’ll use at deployment time (tiled validation changes the problem and can mislead metrics), as outlined in the [YOLO26 + SAHI tiled inference guide](https://docs.ultralytics.com/guides/sahi-tiled-inference/); for the “mostly background tiles” issue, don’t blindly include all negatives—keep all tiles that contain objects and randomly downsample empty tiles per image (start with a small cap like a few empties per positive, then tune based on false positives), and if empties dominate, increasing slice size (and tuning overlap) often helps reduce the 1:30 imbalance while still improving small-object learning per the [tiling notes in our evaluation/fine-tuning insights](https://docs.ultralytics.com/guides/model-evaluation-insights/).

**Reply by cosminlovin7** [reply]
Thanks a lot for the information!

Also, specifically for this scenario (SF+SAHI), do you think keeping the probability to mosaic augmentation to 1.0 would still help the training process? I feel like lowering it a little (maybe 0.5-0.6) especially considering that the objects are very small, and when applying mosaic I expect them to become "microscopic", making it hard for the models to learn anything meaningful.

**Reply by glenn-jocher** [reply, marked as answer]
Yes—your intuition is usually right for SF on already-small 640×640 slices: `mosaic=1.0` can make targets effectively too small after the mosaic + resize step, so I’d try a lower `mosaic` (e.g. `0.3–0.6`) and keep `close_mosaic` enabled (e.g. `10–20`) so the last epochs see “real” tiles; then pick whatever improves your SAHI-style validation metrics. See the docs on [Mosaic augmentation and `close_mosaic`](https://docs.ultralytics.com/guides/yolo-data-augmentation/) for how those knobs behave.

```bash
yolo detect train model=yolo26n.pt data=your.yaml imgsz=640 mosaic=0.5 close_mosaic=20
```

**Reply by cosminlovin7** [reply]
Thank you very much!

**Reply by glenn-jocher** [reply]
Glad it helped—when you get a run you’re happy with, please drop your before/after SAHI-style validation metrics and slice settings here so we can sanity-check the trade-offs, and if you want to systematically tune `mosaic`/`close_mosaic`/overlap without guesswork, the [hyperparameter tuning guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/) is a good next step.
